### ISSUE TITLE
GD.Print(float) on german locale uses comma as decimal separator

### ISSUE THREAD

**Issue by kamisori** [godotengine/godot#43484]
**Godot version:**
3.2.3 mono

**OS/device including version:**
Windows 10 and ubuntu 20 with **locale that uses the comma as decimal separator**

**Issue description:**
I'm using arcadiaGodot to develop and when I ask for a Transform it is returned as such:
((0,4999999, -0,01813666, -0,8658353), (0, 0,9997806, -0,02094242), (0,8660253, 0,01047121, 0,4998902)) - (2,597508, 0,04931428, -8,49967)

**Steps to reproduce:**
GD.Print(0.2); // will print 0,2 on machine using e.g. german locale

[foo.zip](https://github.com/godotengine/godot/files/5533436/foo.zip)

**Comment by Riteo** [comment]
I guess this should be the intended behaviour. What were you expecting?

**Comment by akien-mga** [comment]
Confirmed locally on Linux with `LANG=de LC_ALL=de_DE.UTF-8` with the normal Godot Mono build (without arcadiaGodot).

I remember countless Unity games crashing on Linux in the past due to Mono's "clever" handling of decimal separator which broke expectations of unaware dev serializing data.

IMO if that's something that Mono does, we should see if we can disable it. I don't see why a dev would want their host OS's locale to mess with their cross-platform and cross-language game development.

**Comment by akien-mga** [comment]
Seems we can fix that by setting an invariant culture: https://stackoverflow.com/questions/3135569/how-to-change-symbol-for-decimal-point-in-double-tostring

I label this as `bug` but it's still up for discussion if C# folks think that this is intended behavior in the .NET ecosystem. To me it's confusing and dangerous as mentioned above, but I guess .NET folks had a valid reason to try to force respecting locale settings down to `double.ToString()`.

**Comment by Riteo** [comment]
Yeah, I remember having issues with the decimal separator on .NET some time ago, and setting an invariant culture was the fix.
IMO if a locale-indipendent decimal separator is GDScript's way, I think that it should be also enabled on mono for consistency.

**Comment by aaronfranke** [comment]
> when I ask for a Transform it is returned as such:
> ((0,4999999, -0,01813666, -0,8658353), (0, 0,9997806, -0,02094242), (0,8660253, 0,01047121, 0,4998902)) - (2,597508, 0,04931428, -8,49967)

Probably fixed in this branch: https://github.com/aaronfranke/godot/tree/cs-decimal-culture, let me know if it works for you :)

> Steps to reproduce:
> GD.Print(0.2); // will print 0,2 on machine using e.g. german locale

This is not possible to fix, since `GD.Print` accepts `object`, and `object.ToString()` can't take any additional arguments. If we added overloads for float/double, then `GD.Print(0.2);` would output `0.2`, but then we would still have other problems, such as`GD.Print("num: " + 0.2);` outputting `num: 0,2`.  

If it's deemed better to have the structs behave like core but different from `GD.Print`, vs making the structs behave like `GD.Print` but different from core, I will make the PR. CC @neikeq

Or maybe we can change the default culture for the whole app to InvariantCulture? A quick Google search seems to indicate this is possible, but I'm not sure what's the best way to do it, so I'll let @neikeq investigate this. Even if it is possible, it might cause unexpected results for experienced C# developers who expect the normal culture behavior.

**Comment by Riteo** [comment]
> Even if it is possible, it might cause unexpected results for experienced C# developers who expect the normal culture behavior.

Well, that could be documented and there could be a method for setting the culture to something else.
IMO, since GDScript acts as if it had an invariant culture, mono should follow the same rules.

**Comment by kamisori** [comment]
it's possible to set it for the whole thread:
Thread.CurrentThread.CurrentCulture = System.Globalization.CultureInfo.InvariantCulture;

whether thats desirable or not is another question

**Comment by neikeq** [comment]
>This is not trivial to fix, since `GD.Print` accepts `object`, and `object.ToString()` can't take any additional arguments. If we added overloads for float/double, then `GD.Print(0.2);` would output `0.2`, but then we would still have other problems, such as`GD.Print("num: " + 0.2);` outputting `num: 0,2`. 

I think we could do this instead:

```cs
Convert.ToString(object, CultureInfo.InvariantCulture);
```

**Comment by aaronfranke** [comment]
@neikeq What about the case of `GD.Print("num: " + 0.2);`? That should convert it to `num: 0,2` before we get to `GD.Print`.

**Comment by aaronfranke** [comment]
I think the solution here is to change the culture of the entire project: https://learn.microsoft.com/en-us/dotnet/api/system.globalization.cultureinfo.defaultthreadcurrentculture

**Comment by neikeq** [comment]
> @neikeq What about the case of `GD.Print("num: " + 0.2);`? That would convert it to `num: 0,2` before we get to `GD.Print`.

> I think the solution here is to change the culture of the entire project: https://learn.microsoft.com/en-us/dotnet/api/system.globalization.cultureinfo.defaultthreadcurrentculture

I don't think we should not be changing the culture. That should be up to the project creator. It's not worth it for something harmless like this.

Sorry for not noticing your question until now.

**Comment by akien-mga** [comment]
To me it would make sense that by default, float stringification uses invariant culture, to ensure that a project behaves the same independently of the system it runs on. As I mentioned above, Unity 5 used to have countless bugs in games which had been developed on a system with period as decimal separator, and never tested on a system using comma as decimal separator. We should make sure that, by default, our users are safe from this concern.

We can totally make it configurable though. Maybe a project setting that lets users choose if they want to use the local or invariant culture?

**Comment by raulsntos** [comment]
By default .NET uses the current culture everywhere, this can lead to problems as @akien-mga mentioned which is why today .NET ships with an analyzer ([CA1305](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1305)) that warns you if you are not specifying the culture, pushing you to be more explicit.

For example:

```csharp
int myNumber = 42;
myNumber.ToString();
```

This gets you a warning in the `ToString` invocation that says:

> The behavior of 'int.ToString()' could vary based on the current user's locale settings. Replace this call in 'Program.Main()' with a call to 'int.ToString(IFormatProvider)'.

The solution depends on the developer's intention, if the string is going to be displayed to the user it should use the current culture which follows the user's machine's culture; otherwise, use the invariant culture:

```csharp
myNumber.ToString(CultureInfo.CurrentCulture); // Uses the current culture, this is the default but now we are explicit
myNumber.ToString(CultureInfo.InvariantCulture); // Uses the invariant culture explicitly
```

I agree with @neikeq that changing the project's culture is a bad idea, developers are already able to change it by using existing .NET APIs (see https://learn.microsoft.com/en-us/dotnet/core/extensions/globalization).

Changing the default current culture to the invariant culture will make it impossible for developers to use the user's machine's culture, this would make `CultureInfo.CurrentCulture` and `CultureInfo.InvariantCulture` the same so you can no longer use the current culture in `ToString` when you explicitly want the user's machine's culture.

The culture is also not used just for stringifying values, but also localization (see https://learn.microsoft.com/en-us/dotnet/core/extensions/localization). For example, for users that use `resx` the lookup of a `MessageService.resx` file with the current UI culture being `sr-Cyrl-RS` would be in the following order:

1. _MessageService.sr-Cyrl-RS.resx_
2. _MessageService.sr-Cyrl.resx_
3. _MessageService.sr.resx_
4. _MessageService.resx_

I would maybe support changing the current culture and UI culture to the project's current locale (https://docs.godotengine.org/en/latest/classes/class_translationserver.html#class-translationserver-method-get-locale).

> We can totally make it configurable though. Maybe a project setting that lets users choose if they want to use the local or invariant culture?

I would be OK with making it configurable, but I still feel like the default should remain the user's machine's culture to prevent breaking .NET developers' expectations.

As an example, in ASP.NET you can configure the culture using the `web.config` XML file:

```xml
<configuration>
  <system.web>
    <globalization culture="auto:en-US" uiCulture="fr" />
  </system.web>
</configuration>
```

If we add similar options to project settings, I think that would be fine.

---

We could argue that the default culture used in .NET APIs such as the `ToString` method should have been the invariant culture but that's not going to change now since it would be a huge breaking change, the best we can do now is push developers to be more explicit about the culture they want.

On that note, it may be a good idea to add a `GD.Print` overload that takes `IFormatProvider`.

**Comment by Mickeon** [comment]
Is this issue still relevant nowadays?
